### DISCUSSION TITLE
Cancellation option for Connect & Disconnect

### DISCUSSION THREAD

**Discussion by ArvindRamaswamy** [jchristn/SuperSimpleTcp#185] [Q&A]
Hi, Is there a way to "cancel" a long running "ConnectwithRetries" call from consuming code? Same goes with Disconnect. Reason being, if I initiate a ConnectWithRetries(30000ms) say, and for some reason the connection is not happening fast and say at the end of 10s User wants to "Cancel" the connect operation? Same goes for disconnect. Please help.

**Comment by jchristn** [comment]
Hi @ArvindRamaswamy I'm out on PTO this week but will look into it when I return. If you're interested please make the source changes, submit a PR, and I will integrate. Otherwise I can look at it when I return. 

Cheers, Joel

**Comment by k7mck** [comment]
I had the same question.  I wanted my tcpClient connection on it's own thread, too - and I wanted to update my UI to show connection attempt progress.
I solved this by creating two tasks: 1. creates the tcpClient and tcpClientThread 2. updates the UI, based on how much time has passed out of the retries time.

`internal static async Task<bool> CreateTcpClientConnection(CancellationToken cts)`

Inside CreateTcpClientConnection, I initialize some stuff:
`Exception tcpClientThreadException = null;
tcpClientThread = new Thread(() => SafeExecute(() => tcpClient.ConnectWithRetries(connectRetriesTimeMs), out tcpClientThreadException));
tcpClientThread.IsBackground = true;
tcpClientThread.Start();`

Within CreateTcpClientConnection, right after the tcpClient is created and the tcpClientThread is started, I AWAIT my second task using the same cancellation token (cts):
`await Task.Run(() => ConnectionUpdate(progress, cts),cts);`

In my app, the user can't do anything until the tcpClient is connected, so it's useful to see the connection is "in progress".
Within both tasks, there are checks on the cancellation token to, respectively: A. immediately stop updating the UI progress. B. Disconnect() and Dispose() of the tcpClient.

As soon the await Task.Run returns, I check the cancellation token. That's it - if it was canceled, I Disconnect/Dispose of the tcpClient.

Example: I create a tcpClient with ConnectWithRetries value of 20 seconds, and then kick off my thread.  Within 5 seconds the user hits a Cancel button which fires the CancellationTokenSource.Cancel(), and in my tasks, the source's associated tokens now have token.IsCancellationRequested as True, and I can act accordingly. My awaited Task.Run returns immediately, and I wipe the tcpClient - I don't have to wait for the tcpClient to finish it's retries - poof, it's gone.

You could def write this differently - but use cancellation sources and tokens (ConnectWithRetries uses them internally).

***I had this method to make sure thread exceptions get caught:
`private static void SafeExecute(Action method, out Exception exception)
{
            exception = null;
            try{method.Invoke();}
            catch (Exception ex)
            {exception = ex;}
}`

***This is my method for updating my UI's connection status label - every second it updates the UI progress:
`private static void ConnectionUpdate(IProgress<string> progress, CancellationToken cts)
        {
            int waitCount = 0;
            //allow three seconds for the connection to succeed
            while (!tcpClient.IsConnected && waitCount <= (connectRetriesTimeMs / 1000))
            {
                //immediately return if user has cancelled the connection attempt
                if (cts.IsCancellationRequested) { return; }
                string dots = new string('*', waitCount);
                progress.Report(dots + " Connecting to service " + dots);
                System.Threading.Thread.Sleep(1000);//wait one second
                waitCount++;
            }
            return;
}`
